### ISSUE TITLE
ComplexBatchNorm在训练过程中内存占用不断升高

### ISSUE THREAD

**Issue by RuqiaoLiu** [huyanxin/DeepComplexCRN#20]
您好：
我发现把ComplexBatchNorm中的track_running_stats设置为True，在训练过程中的内存占用会不断上升，最终导致内存爆炸。请问您在训练DCCRN时没有出现该问题吗？
谢谢；
祝好！

**Comment by changingBin** [comment]
我也遇到这个问题了 排查了好久才发现的

**Comment by RuqiaoLiu** [comment]
嗯，建议使用speechBrain里写的复BN。

--------------原始邮件--------------
发件人："changingBin ***@***.***&gt;;
发送时间：2022年1月16日(星期天) 下午3:05
收件人："huyanxin/DeepComplexCRN" ***@***.***&gt;;
抄送："刘汝桥 ***@***.***&gt;;"Author ***@***.***&gt;;
主题：Re: [huyanxin/DeepComplexCRN] ComplexBatchNorm在训练过程中内存占用不断升高 (Issue #20)
-----------------------------------

 

 
我也遇到这个问题了 排查了好久才发现的
 
—
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by victorVoice** [comment]
https://discuss.pytorch.org/t/how-does-batchnorm-keeps-track-of-running-mean/40084/15 可以参考这个来做修改，应该也是再计算running_mean, runing_var的时候lerp把mean，var也加入计算图里面了

**Comment by FragrantRookie** [comment]
我也遇到了这个问题，请问你解决了吗？

**Comment by victorVoice** [comment]
> 
complexnn 文件里
        if self.training and self.track_running_stats:
            self.num_batches_tracked += 1
            if self.momentum is None:  # use cumulative moving average
                exponential_average_factor = 1.0 / self.num_batches_tracked.item()
            else:  # use exponential moving average
                exponential_average_factor = self.momentum
.item() 加一个detach就可以

**Comment by TungyuYoung** [comment]
> > 
> 
> complexnn 文件里 if self.training and self.track_running_stats: self.num_batches_tracked += 1 if self.momentum is None: # use cumulative moving average exponential_average_factor = 1.0 / self.num_batches_tracked.item() else: # use exponential moving average exponential_average_factor = self.momentum .item() 加一个detach就可以

您好，我加了detach()之后似乎内存还在继续增长，因此我在前面使用了torch.no_grad()，但问题仍然没有得到解决。请问您还有更好的建议吗
